sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,307 @@
+	lastUpdateTime                    time.Time
+	enableMultipleWriteLocations      bool
+	unavailableLocationExpirationTime time.Duration
+}
+
+func newLocationCache(prefLocations []string, defaultEndpt url.URL) *locationCache {
+	return &locationCache{
+		defaultEndpt:                      defaultEndpt,
+		locationInfo:                      *newDbAcctLocationsInfo(prefLocations, defaultEndpt),
+		locationUnavailabilityInfoMap:     make(map[url.URL]locationUnavailabilityInfo),
+		unavailableLocationExpirationTime: defaultExpirationTime,
+	}
+}
+
+func (lc *locationCache) update(writeLocations []AcctRegion, readLocations []AcctRegion, prefList []string, enableMultipleWriteLocations bool) error {
+	lc.rwMutex.RLock()
+	nextLoc := copyDbAcctLocationsInfo(lc.locationInfo)
+	if prefList != nil {
+		nextLoc.prefLocations = prefList
+	}
+	lc.enableMultipleWriteLocations = enableMultipleWriteLocations
+	lc.refreshStaleEndpts()
+	if readLocations != nil {
+		availReadEndptsByLocation, availReadLocations, err := getEndptsByLocation(readLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availReadEndptsByLocation = availReadEndptsByLocation
+		nextLoc.availReadLocations = availReadLocations
+	}
+
+	if writeLocations != nil {
+		availWriteEndptsByLocation, availWriteLocations, err := getEndptsByLocation(writeLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availWriteEndptsByLocation = availWriteEndptsByLocation
+		nextLoc.availWriteLocations = availWriteLocations
+	}
+
+	nextLoc.writeEndpts = lc.getPrefAvailableEndpts(nextLoc.availWriteEndptsByLocation, nextLoc.availWriteLocations, write, lc.defaultEndpt)
+	nextLoc.readEndpts = lc.getPrefAvailableEndpts(nextLoc.availReadEndptsByLocation, nextLoc.availReadLocations, read, nextLoc.writeEndpts[0])
+	lc.lastUpdateTime = time.Now()
+	lc.rwMutex.RUnlock()
+	lc.rwMutex.Lock()
+	lc.locationInfo = nextLoc
+	lc.rwMutex.Unlock()
+	return nil
+}
+
+func (lc *locationCache) readEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.readEndpts, nil
+}
+
+func (lc *locationCache) writeEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.writeEndpts, nil
+}
+
+func (lc *locationCache) getLocation(endpoint url.URL) string {
+	firstLoc := ""
+	lc.rwMutex.RLock()
+	defer lc.rwMutex.RUnlock()
+	// TODO: Find workaround for firstLoc, maps are unordered
```

Comment:
Would be ideal to see if there is a shorter/cleaner way to do this, the code looks like it does the same as 
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs#L131-L143 but it took me a bit to digest

Reply:
I understand, let me know if you have any direct suggestions, but this was the way to work around the fact that there's no direct way to enumerate a map without a for-range that I know of, which is needed to imitate the FirstOrDefault in C#.

Reply:
No suggestions, my lack of Go experience shows :)